File: src/markov/MarkovChainGen3.java
package markov;

import java.util.List;

public class MarkovChainGen3 extends MarkovChainGen2 {

    private MarkovWordGraph wgTriples;

    protected String thirdLastWord;

    public MarkovChainGen3() {
        super();
        wgTriples = new MarkovWordGraph(START_SIGNAL, END_SIGNAL);

        thirdLastWord = null;
    }

    public MarkovChainGen3(String start, String end) {
        super();
        wgTriples = new MarkovWordGraph(start, end);

        thirdLastWord = null;
    }

    public MarkovChainGen3(String filename){
        this();
        train(filename);
    }


    @Override
    public void train(String filename) {
        super.train(filename);
        wgTriples.processFileGen(FILEPATH + filename, 3);
    }

    @Override
    public List<String> getNextWords() {
        var previousOrder = super.getNextWords();

        if (thirdLastWord == null)
            return previousOrder;

        var strToFind = "[" + thirdLastWord + ", " + getSecondLastWord() + ", " + getLastWord() + "]";

        return getStrings(previousOrder, strToFind, wgTriples);
    }

    @Override
    public void updateMemory() {
        thirdLastWord = getSecondLastWord();
        super.updateMemory();
    }

    @Override
    public String generateSentence() {
        thirdLastWord = null;

        return super.generateSentence();
    }
}

File: src/markov_kt/MarkovChainGen3.kt
package markov_kt

import markov.MarkovChainGen2
import markov_kt.MarkovChainGen2.Companion.getStrings

class MarkovChainGen3 : MarkovChainGen2 {
    var wgTriples: MarkovWordGraph
    protected var thirdLastWord: String?

    constructor() : super() {
        wgTriples = MarkovWordGraph(START_SIGNAL, END_SIGNAL)
        thirdLastWord = null
    }

    constructor(start: String, end: String?) : super() {
        wgTriples = MarkovWordGraph(start, end)
        thirdLastWord = null
    }

    constructor(filename: String, start: String, end: String?) : super() {
        wgTriples = MarkovWordGraph(start, end)
        thirdLastWord = null
        train(filename)
    }

    constructor(filename: String) : this() {
        train(filename)
    }

    override fun train(filename: String) {
        super.train(filename)
        wgTriples.processFileGen(FILEPATH + filename, 3)
    }

    override fun getNextWords(): List<String> {
        val previousOrder = super.getNextWords()
        if (thirdLastWord == null) return previousOrder
        val strToFind = "[" + thirdLastWord + ", " + secondLastWord + ", " + getLastWord() + "]"
        return getStrings(previousOrder, strToFind, wgTriples)
    }

    override fun updateMemory() {
        thirdLastWord = secondLastWord
        super.updateMemory()
    }

    override fun generateSentence(): String {
        thirdLastWord = null
        return super.generateSentence()
    }
}